docs/src/reference/experimental/loop-contracts.md

```diff
@@ -0,0 +1,109 @@
+Therefore, the invariant must hold after the loop execution for any number of iterations. The invariant, together with the negation of the loop guard,
+must be sufficient to establish subsequent assertions. If it is not, the abstraction is too imprecise and the user must supply a stronger invariant.
+
+To illustrate the key idea, we show how Kani abstract the loop in `simple_loop_with_loop_contracts` as a non-loop block:
+``` Rust
+assert!(x >= 1) // check loop invariant for the base case.
+x = kani:any();
+assume!(x >= 1);
+if x > 1{
+    // proof path 1:
+    //   both loop guard and loop invariant are satisfied.
+    x = x - 1;
+    assert!(x >= 1); // check that loop invariant is inductive.
+    assume!(false) // kill this proof path.
+}
+// proof path 2:
+//   loop invariant is satisfied and loop guard is violated.
+assert!(x == 1);
+```
+That is, we assume that we are in an arbitrary iteration after checking that the loop invariant holds for the base case. With the inductive hypothesis (`assume!(x >= 1);`),
+we will either enter the loop (proof path 1) and leave the loop (proof path 2). We prove the two paths separately by killing the path 1 with `assume!(false);`.
+Note that all assertions after `assume!(false)` will be ignored as `false => p` can be deduced as `true` for any `p`.
+
+In the proof path 1, we prove properties inside the loop and at last check that the loop invariant is inductive.
+
+In the proof path 2, we prove properties after leaving the loop. As we leave the loop only when the loop guard is violated, the post condition of the loop can be expressed as
+`!guard && inv`, which is `x <= 1 && x >= 1` in the example. The postcondition implies `x == 1`---the property we want to prove at the end of `simple_loop_with_loop_contracts`.
+
+
+## Limitations
+
+Loop contracts comes with the following limitations.
+
+1. Besides `while` loops, there are three other kinds of loops that we don't support loop contracts for: [`loop` loops](https://doc.rust-lang.org/reference/expressions/loop-expr.html#infinite-loops)
+   , [`while let` loops](https://doc.rust-lang.org/reference/expressions/loop-expr.html#predicate-pattern-loops), and [`for` loops](https://doc.rust-lang.org/reference/expressions/loop-expr.html#iterator-loops).
+2. Kani infer *loop modifies* with alias analysis. Loop modifies are those variables we assume to be arbitrary in the inductive hypothesis, and should cover all memory locations that are written to during the execution of the loops. Proof will fail if the inferred loop modifies miss some targets written in the loops.
+   We observes this happen when some fields of structs are modified by some other functions called in the loops.
+3. Kani doesn't check if a loop will always terminate in proofs with loop contracts. So it could be that some properties are proved successfully with Kani but actually are unreachable due to the non-termination of some loops.
```

Comment:
Will they be reported as [unreachable](https://model-checking.github.io/kani/verification-results.html#check-results) in this case? If so, we should point that out.

Reply:
No, they won't be reported as unreachable.

Reply:
Do you know why they aren't reported as unreachable?

Reply:
Yes. We execute only one non-deterministic iteration of the loop when applying loop contracts. So won't know if there will be infinite number of iteration to be executed. In CBMC, we have decreases clauses to check if there are some quantities decreasing along the execution of the loop. Loops will terminate if such quantities exist. However, decreases clauses are not implemented in Kani yet.

Reply:
Got it, thanks.